src/dbnode/storage/shard.go

```diff
@@ -2660,74 +2661,216 @@ func (s *dbShard) Repair(
+		if !closed {
+			if err := readerIter.Close(); err != nil {
+				// NB: log the error on ungraceful exit.
+				s.logger.Error("could not close read iterator on error", zap.Error(err))
+			}
+		}
+	}()
+
+	encoder := s.opts.EncoderPool().Get()
+	defer encoder.Close()
+	encoder.Reset(opts.Start, 0, targetSchemaDescr)
+
+	latestTargetVolume, err := s.LatestVolume(opts.Start)
+	if err != nil {
+		return 0, err
+	}
+
+	nextVolume := latestTargetVolume + 1
+	writer, err := fs.NewStreamingWriter(
+		fs.StreamingWriterOptions{
+			NamespaceID:         s.namespace.ID(),
+			ShardID:             s.ID(),
+			Options:             s.opts.CommitLogOptions().FilesystemOptions(),
+			BlockStart:          opts.Start,
+			BlockSize:           s.namespace.Options().RetentionOptions().BlockSize(),
+			VolumeIndex:         nextVolume,
+			PlannedRecordsCount: uint(maxEntries),
 		},
-		FileSetType: persist.FileSetFlushType,
-		//TODO add after streaming supported - OrderByIndex: true
+	)
+	if err != nil {
+		return 0, err
 	}
-	if err := reader.Open(openOpts); err != nil {
+
+	if err := writer.Open(); err != nil {
 		return 0, err
 	}
-	defer reader.Close()
 
-	encodingOpts := encoding.NewOptions().SetBytesPool(s.opts.BytesPool())
-	bytesReader := bytes.NewReader(nil)
-	dataPointIter := m3tsz.NewReaderIterator(bytesReader, m3tsz.DefaultIntOptimizationEnabled, encodingOpts)
-	var lastWriteError error
-	var processedBlockCount int64
+	var (
+		processedTileCount atomic.Int64
+		multiErr           xerrors.MultiError
+		dp                 ts.Datapoint
+		segmentCapacity    int
+		writerData         = make([][]byte, 2)
+	)
 
-	for {
-		id, tags, data, _, err := reader.Read()
-		if err == io.EOF {
-			break
+	READER:
+	for readerIter.Next() {
+		seriesIter, id, encodedTags := readerIter.Current()
+
+		for seriesIter.Next() {
+			frame := seriesIter.Current()
+			unit, singleUnit := frame.Units().SingleValue()
+			annotation, singleAnnotation := frame.Annotations().SingleValue()
+
+			if frameValues := frame.Values(); len(frameValues) > 0 {
+				lastIdx := len(frameValues) - 1
+
+				value := frameValues[lastIdx]
+				timestamp := frame.Timestamps()[lastIdx]
+
+				if !singleUnit {
+					// TODO: what happens if unit has changed mid-tile?
+					// Write early and then do the remaining values separately?
+					if lastIdx < len(frame.Units().Values()) {
+						unit = frame.Units().Values()[lastIdx]
+					}
+				}
+
+				if !singleAnnotation {
+					// TODO: what happens if annotation has changed mid-tile?
+					// Write early and then do the remaining values separately?
+					if lastIdx < len(frame.Annotations().Values()) {
+						annotation = frame.Annotations().Values()[lastIdx]
+					}
+				}
+
+				dp.Timestamp = timestamp
+				dp.Value = value
+				err = encoder.Encode(dp, unit, annotation)
+				if err != nil {
+					s.metrics.largeTilesWriteErrors.Inc(1)
+					multiErr = multiErr.Add(err)
+					encoder.Reset(opts.Start, segmentCapacity, targetSchemaDescr)
```

Comment:
nit: Maybe call Reset just before entering this loop? (rather than explicitly on continue)

That way if we add other "continue" statements we don't need to explicitly reset and it can just be a "setup" thing we do per-series where we can do other "prepare to write series" tasks.

Reply:
This would mean a double reset on the happy path (which would then become `Reset` + `DiscardReset`). I cannot simply replace `DiscardReset` with `Discard` for the happy path because `Discard` returns the `encoder` itself to the pool making it non-reusable (I'll update the comments on `encoder` methods to make this more clear). So I don't see how I could make such change nicely without modifying the API of `encoder`.
However, this whole question might just go away, see my answer on the follow up suggestion.